witchcraft/wdebug/resource.go

```diff
@@ -0,0 +1,139 @@
+	headerKeySafeLoggable = "Safe-Loggable"
+)
+
+type DiagnosticType string
+
+type debugResource struct {
+	SharedSecret refreshable.String
+}
+
+func RegisterRoute(router wrouter.Router, sharedSecret refreshable.String) error {
+	debugResource := debugResource{SharedSecret: sharedSecret}
+	if err := wresource.New("witchcraftdebugservice", router).
+		Get("GetDiagnostic", "/debug/diagnostic/{diagnosticType}",
+			httpserver.NewJSONHandler(debugResource.ServeHTTP, httpserver.StatusCodeMapper, httpserver.ErrHandler),
+		); err != nil {
+		return werror.Wrap(err, "failed to register WitchcraftDebugService")
+	}
+	return nil
+}
+
+func (r *debugResource) ServeHTTP(rw http.ResponseWriter, req *http.Request) error {
+	ctx := req.Context()
+	authHeader, err := httpserver.ParseBearerTokenHeader(req)
+	if err != nil {
+		return errors.WrapWithNewError(err, errors.DefaultPermissionDenied)
+	}
+	if authHeader != r.SharedSecret.CurrentString() {
+		return errors.NewPermissionDenied()
+	}
+
+	pathParams := wrouter.PathParams(req)
+	if pathParams == nil {
+		return werror.WrapWithContextParams(ctx, errors.NewInternal(), "path params not found on request: ensure this endpoint is registered with wrouter")
+	}
+	diagnosticTypeStr, ok := pathParams["diagnosticType"]
+	if !ok {
+		return werror.WrapWithContextParams(ctx, errors.NewInvalidArgument(), "path param not present", werror.SafeParam("pathParamName", "diagnosticType"))
+	}
+	diagnosticType := DiagnosticType(diagnosticTypeStr)
+
+	return r.writeDiagnostic(ctx, diagnosticType, rw)
+}
+
+func (r *debugResource) writeDiagnostic(ctx context.Context, diagnosticType DiagnosticType, rw http.ResponseWriter) error {
+	switch diagnosticType {
+	case "threaddump.v1":
+		return r.getThreadDumpV1(ctx, rw)
+	case "go.profile.cpu.v1":
+		return r.getCPUProfileV1(ctx, rw)
+	case "go.profile.heap.v1":
+		return r.getHeapProfileV1(ctx, rw)
+	case "go.profile.allocs.v1":
+		return r.getAllocsProfileV1(ctx, rw)
+	default:
+		return errors.WrapWithInvalidArgument(werror.ErrorWithContextParams(ctx, "unsupported diagnosticType", werror.SafeParam("diagnosticType", diagnosticType)))
+	}
+}
+
+func (r *debugResource) getThreadDumpV1(ctx context.Context, rw http.ResponseWriter) error {
+	var buf bytes.Buffer
+	_ = pprof.Lookup("goroutine").WriteTo(&buf, 2) // bytes.Buffer's Write never returns an error, so we swallow it
+	threads := diag1log.ThreadDumpV1FromGoroutines(buf.Bytes())
+
+	rw.Header().Set(headerKeyContentType, codecs.JSON.ContentType())
+	rw.Header().Set(headerKeySafeLoggable, "true")
+	return codecs.JSON.Encode(rw, threads)
+}
+
+func (r *debugResource) getCPUProfileV1(ctx context.Context, rw http.ResponseWriter) error {
+	// If diagnostics ever support parameters, this can be configurable
+	const defaultDuration = 30 * time.Second
+
+	rw.Header().Set(headerKeyContentType, codecs.Binary.ContentType())
+	rw.Header().Set(headerKeySafeLoggable, "true")
+
+	if err := pprof.StartCPUProfile(rw); err != nil {
```

Comment:
Do you think it's worth storing any state to detect when we're in the middle of a profile already, so that we can return a more informed error. I noticed this will return an error if one is already in flight.

Reply:
I'd kind of prefer to let the std lib do its thing rather than try to reimplement it with something like a mutex. Right now you'll get a 500 response with a logged error like `failed to start CPU profile: cpu profiling already in use`. 

Updated to make sure that message makes it into the conjure error and called it a Conflict (409)